docs/rfc/0696-p2p-node.adoc

```diff
@@ -0,0 +1,318 @@
+
+==== Announcements
+
+Until superseded by graftier ways, currently the main means of dissemination is
+through gossiping new interesting `Have`s. This is achieved by maintainig a
+persisted view on the list of identities and their current refs. During
+periodic runs a delta is created which serves as the basis of the announcements.
+In turn that list is made known to actively connected peers via the core
+protocols gossip.
+
+==== Requests
+
+The core protocol implicitly tracks projects after replication, and explicitely
+with the `track` operation for remotes in the context of a `Urn`. To enable the
+same approach as for announcements where offline changes to the storage can be
+picked up at any point the node is running and online, this subroutine will
+maintain a persisted view and periodically (potentially provoked by a request
+over the <<API>>) build the delta for new tracked `Urn`s and remotes in their
+context.
+
+Note: A blocking issue for that is
+https://github.com/radicle-dev/radicle-link/issues/141[#141]
+
+==== Tracking
+
+Given the ability to configure a set of urn, peers, both or anything observed
+from connected peers, the node will automatically track and replicate. This
+strategy is currently in use in the seed node and SHALL be preserved.
+
+== API
+
+=== IPC
+
+All communication with daemon processes SHALL occur over UNIX domain sockets in
+`SOCK_STREAM` mode. The socket files MUST be stored in a directory only
+accessible by the logged-in user (typically `$XDG_RUNTIME_DIR`), and have
+permissions `0700` by default. Per-service socket paths are considered "well
+known".
+
+RPC calls over those sockets use <<cbor, CBOR>> for their payload encoding. As
+incremental decoders are not available on all platforms, CBOR-encoded messages
+shall be prepended by their length in bytes, encoded as a 32-bit unsigned
+integer in network byte order.
+
+RPC messages are wrapped in either a `request` or `response` envelope structure
+as defined below:
+
+[source,cddl]
+----
+request = [
+    request-headers,
+    ? payload: bstr,
+]
+response = ok / error
+ok = [
+    response-headers,
+    ? payload: bstr,
+]
+error = [
+    response-headers,
+    code: uint,
+    ? message: tstr,
+]
+request-headers = {
+    ua: client-id,
+    rq: request-id,
+    ? token: token,
+}
+response-headers = {
+    rq: request-id,
+}
+; Unambiguous, human-readable string identifying the client application. Mainly
+; for diagnostic purposes. Example: "radicle-link-cli/v1.2+deaf"
+client-id: tstr .size (4..16)
+; Request identifier, choosen by the client. The responder includes the
+; client-supplied value in the response, enabling request pipelining.
+;
+; Note that streaming / multi-valued responses may include the same id in
+; several response messages.
+request-id: bstr .size (4..16)
+
+; Placeholder for future one-time-token support.
+token: bstr
+----
+
+
+=== Request/Response payloads
```

Comment:
> Should the `Requests` action also have a `request`? Or put another way, why doesn't it make the cut?

Took me a couple of reads to unpack these sentences. ^^

It didn't make the cut so far, because the main purpose of the API was to introspect. Which is not to say that using it for other means, i.e. more interactive use-cases is unattainable. What would the `request` request - yo dawg we need to find a better name - do tho?

> I'm also wondering if it might be worth having a request for announcements. I think we might have discussed this offline -- I think we landed on the daemon running announcements when it would come online, however, I think it might be useful to have an explicit mechanism to announce. It could be useful for testing, for example.

Similar question as above what would the request/response look like here, and more importantly what's the interplay with the announcement protocol? Given that it's tailored to only announce deltas, would the request trigger announcements of everything?

Reply:
> What would the request request - yo dawg we need to find a better name - do tho?

So I think this is related to the discussion here https://github.com/radicle-dev/radicle-link/pull/698/files#r660397972. In general, I think it would be useful to have explicit mechanisms for announcements and replication. At least for debugging in a test net scenario it will be immensely helpful. I'm envisioning a scenario where I'm running a test net and announcements aren't working properly and, for example, the announcement protocol isn't going to announce again because it's based on deltas. If I have a mechanism to announce a SHA manually for a Urn then I have more ample opportunity to correlate that explicit call with the logs/tracing.

Now, I realise that the downside is that this opens the door for developers (like you and me in our naivete) to possibly abuse these mechanisms and flood announcements in a shotgun fashion. But at least we could test the rate-limiting then ;)

What do yourself and others think?

Reply:
@FintanH Would it be sensible to defer tooling like the one you are asking for when we look at the APIs for *authorised personnel* (similar to how we settled to do introspection on top of stabilised p2p node and cli)?